Improve Passive AI (corrected commit content)

Gaelmare opened the pull request:
Tweaked distances and avoidance speeds for avoid AI
**They're not happy around humans anymore**
Pulled most prey AI tasks into common functions on EntityAnimal.
Added swimmingAI to cows
Eating grass AI to all livestock

Incorporated all other TFC specific prey animal AI (not vanilla derived entities).

Is it really ok to have 9 EntityAIAvoidEntity tasks for the all the predators, plus EntityPlayer, and EntityMob?

Copied from #1128 without merge/rebase wierdness

Gaelmare committed:
Cleared IPredator searching out.

Gaelmare committed:
AI Improvements to match classic usage and make passive mobs more elusive. Partially fixes #1032.
Makes AIAvoidEntity scan for IPredator, which isn't an Entity, but all Predators are. This should improve performance.

Gaelmare committed:
Cleared IPredator searching out.

Gaelmare committed:
AI Improvements to match classic usage and make passive mobs more elusive. Partially fixes #1032.
Makes AIAvoidEntity scan for IPredator, which isn't an Entity, but all Predators are. This should improve performance.

Gaelmare committed:
Changes as requested from review, incorporated common prey animal AIs with varied speeds.

Gaelmare committed:
Merge branch 'improveai' of https://github.com/Gaelmare/TerraFirmaCraft into improveai

 Conflicts:
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityChickenTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityCowTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityDeerTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityPheasantTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityPigTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntityRabbitTFC.java
	src/main/java/net/dries007/tfc/objects/entity/animal/EntitySheepTFC.java

Gaelmare committed:
Code Style Fixes

DisasterMoo commented on src/main/java/net/dries007/tfc/objects/entity/animal/EntityAnimalTFC.java:
```diff
@@ -412,4 +417,39 @@ public void onLivingUpdate()
+        entity.tasks.addTask(0, new EntityAISwimming(entity));
+        entity.tasks.addTask(1, new EntityAIPanic(entity, 1.4D * speedMult));
+        //space for livestock AIMate and AITempt
+        entity.tasks.addTask(4, new EntityAIAvoidEntity<>(entity, EntityPlayer.class, 12.0F, farSpeed, nearSpeed));
```
One minor nitpick: Did you test what this causes to familiarized animals? Because i think this will make it run from players regardless of familiarity stat.

Suggestion: Create an `EntityAIAvoidPlayer` extension from vanilla's with the player class already set and add the familiarity check. Anything non-zero sholdn't run from you since it would means it's not scared of you anymore.

Gaelmare replied:
Done. Also applied it to Wolves!

DisasterMoo commented on src/main/java/net/dries007/tfc/api/types/ILivestockAI.java:
```diff
@@ -0,0 +1,57 @@
+ * Does apply to at least Pig, Cow, Chicken, Sheep, Goat, Duck, Alpaca,
+ * Used only in TFC worlds.
+ */
+public interface ILivestockAI extends ILivestock
```
Since you moved logic to static methods in `EntityAnimal`, missing cleanup

DisasterMoo reviewed (changes requested):
Missing cleanup and a minor issue, see below.

agilmore2 committed:
PR Review changes.
New AITamableAvoidPlayer will avoid players unless familiarized at least a bit.
Slick Predicate usage to avoid all IPredators in one AI task
Wolves also avoid players

Gaelmare commented:
Ready for review, also added AvoidTamablePlayer to wolves!

DisasterMoo commented on src/main/java/net/dries007/tfc/objects/entity/ai/EntityAITamableAvoidPlayer.java:
```diff
@@ -0,0 +1,29 @@
+import net.dries007.tfc.api.types.IAnimalTFC;
+
+//AI to avoid players unless familiarized. Must be applied to EntityAnimal that implements IAnimalTFC
+public class EntityAITamableAvoidPlayer extends EntityAIAvoidEntity
+{
+    public EntityAITamableAvoidPlayer(EntityAnimal entityIn, Class classToAvoidIn, float avoidDistanceIn, double farSpeedIn, double nearSpeedIn)
+    {
+        super(entityIn, classToAvoidIn, avoidDistanceIn, farSpeedIn, nearSpeedIn);
+        if (!(entityIn instanceof IAnimalTFC))
+        {
+            throw new AssertionError();
+        }
+    }
```
```suggestion
public class EntityAITamableAvoidPlayer<T extends EntityCreature & IAnimalTFC> extends EntityAIAvoidEntity<EntityPlayer>
{
    public EntityAITamableAvoidPlayer(T entityIn, float avoidDistanceIn, double farSpeedIn, double nearSpeedIn)
    {
        super(entityIn, EntityPlayer.class, avoidDistanceIn, farSpeedIn, nearSpeedIn);
    }
```

Gaelmare replied:
Done. Also needed to consider begging in wolf avoidance.

agilmore2 committed:
PR Review changes II
Template usage FTW
Use begging flag to shut off AvoidAI for wolves

agilmore2 committed:
Tune common speeds for avoid AIs.
